### PR TITLE
[ctx_prof] Insert the ctx prof flattener after the module inliner

### PR DESCRIPTION
**Pull request by mtrofin**
This patch enables experimenting with the contextual profile. ICP is currently disabled in this case - will reenable it subsequently. Also subsequently the inline cost model / decision making would be updated to be context-aware. Right now, this just achieves "complete use" of the profile, in that it's ingested, maintained, and sunk to a flat profile when not needed anymore.

Issue [#89287](https://github.com/llvm/llvm-project/issues/89287)

### CODE REVIEW

**Comment by mtrofin** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/107499?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#107499** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/107499?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#107329** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/107329?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#107463** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/107463?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 1]
```diff
@@ -1744,11 +1749,14 @@ ModulePassManager PassBuilder::buildThinLTODefaultPipeline(
     MPM.addPass(GlobalDCEPass());
     return MPM;
   }
-
-  // Add the core simplification pipeline.
-  MPM.addPass(buildModuleSimplificationPipeline(
-      Level, ThinOrFullLTOPhase::ThinLTOPostLink));
-
+  if (!UseCtxProfile.empty()) {
+    MPM.addPass(
```
Comment (david-xl) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:1753]: this change reduces the number of simplification passes (pre and post inliner) run. What is the reason?

Comment (david-xl) [reply llvm/lib/Passes/PassBuilderPipelines.cpp:1753]: OR the intention is to do a quick e2e integration?

**Review by david-xl — COMMENTED** [review]
what is problem with ICP?

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 2]
```diff
@@ -1017,6 +1017,11 @@ PassBuilder::buildModuleInlinerPipeline(OptimizationLevel Level,
   IP.EnableDeferral = false;
 
   MPM.addPass(ModuleInlinerPass(IP, UseInlineAdvisor, Phase));
+  if (!UseCtxProfile.empty()) {
+    MPM.addPass(GlobalOptPass());
+    MPM.addPass(GlobalDCEPass());
+    MPM.addPass(PGOCtxProfFlatteningPass());
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Passes/PassBuilderPipelines.cpp:1023]: yes. thanks.

File: llvm/lib/Passes/PassBuilderPipelines.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Passes/PassBuilderPipelines.cpp:1753]: Right now yes, it's just to integrate. The next step plan is to run some of the module simplification after all IPO and flattened profile lowering happens, but I want to see first what the more meaningful to do that is - special case through `buildModuleSimplificationPipeline` or have them spelled out separately. I'll evaluate this experimentally.

**Comment by mtrofin** [comment]
> what is problem with ICP?

There's no problem, but I need to add a criteria for when to ICP, and that would increase the size of this patch. That's why it's punted.

**Review by david-xl — APPROVED** [review]
OK as an incremental step for an experimental feature.

**Comment by mtrofin** [comment]
### Merge activity

* **Sep 9, 9:15 PM EDT**: @mtrofin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/107499).

**Pull request merged**